### ISSUE TITLE
Sporadic tls: bad certificate Issue

### ISSUE THREAD

**Issue by therealsamlin** [hashicorp/vault-k8s#141]
Env Versions:

**GKE 1.14.10-gke.36**
**vault**:1.2.1
**agent_inject_vault_image**: vault:1.3.2
**vault-agent-injector**: hashicorp/vault-k8s:0.3.0

---

I'm seeing sporadic tls: bad certificate error log entries in the **vault-agent-injector** container. This is causing vault-agent containers not being injected into our pods and causing apps to crash. Wonder if anyone has seen this or better yet could point out my mistakes 🙏🏻

vault-agent-injector.log
```
2020-06-02T05:47:13.961421898Z 2020-06-02T05:47:13.953Z [INFO]  handler: Starting handler..
2020-06-02T05:47:13.961490627Z Listening on ":8080"...
2020-06-02T05:47:13.980111277Z Updated certificate bundle received. Updating certs...
2020-06-02T06:02:49.031206881Z 2020-06-02T06:02:49.030Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:16:58.772934263Z 2020-06-02T06:16:58.771Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:19:12.222721585Z 2020-06-02T06:19:12.217Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:29:45.604690396Z 2020-06-02T06:29:45.603Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:38:44.372917575Z 2020-06-02T06:38:44.372Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:59:32.662445342Z 2020-06-02T06:59:32.662Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:59:33.312342263Z 2020-06-02T06:59:33.312Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:59:33.325205594Z 2020-06-02T06:59:33.325Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:59:47.241623679Z 2020-06-02T06:59:47.239Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T06:59:47.415966009Z 2020-06-02T06:59:47.413Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:00:21.976433396Z 2020-06-02T07:00:21.966Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:00:22.289668763Z 2020-06-02T07:00:22.219Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:01:50.033803963Z 2020-06-02T07:01:50.031Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:02:27.804240219Z 2020-06-02T07:02:27.804Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:02:27.831802107Z 2020-06-02T07:02:27.831Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:02:28.618867665Z 2020-06-02T07:02:28.618Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:07:30.679237828Z 2020-06-02T07:07:30.676Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:07:31.055053655Z 2020-06-02T07:07:31.034Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:07:36.267515821Z 2020-06-02T07:07:36.267Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:07:36.279054864Z 2020-06-02T07:07:36.278Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:07:52.096649591Z 2020-06-02T07:07:52.079Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:07:52.096678807Z 2020-06-02T07:07:52.091Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:08:04.433026731Z 2020-06-02T07:08:04.432Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:08:10.692660442Z 2020-06-02T07:08:10.689Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:18:40.396171467Z 2020-06-02T07:18:40.395Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:18:40.578473213Z 2020-06-02T07:18:40.578Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:18:40.919113638Z 2020-06-02T07:18:40.918Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:20:14.334225402Z 2020-06-02T07:20:14.319Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:28:07.110445901Z 2020-06-02T07:28:07.109Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T07:28:30.00637003Z 2020-06-02T07:28:30.004Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T09:10:52.530310714Z 2020-06-02T09:10:52.530Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T09:12:56.065020085Z 2020-06-02T09:12:56.056Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T09:13:35.498441485Z 2020-06-02T09:13:35.493Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T09:13:35.502775497Z 2020-06-02T09:13:35.501Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T10:30:28.366852121Z 2020-06-02T10:30:28.366Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T10:30:28.588449658Z 2020-06-02T10:30:28.588Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T11:37:04.189491886Z 2020-06-02T11:37:04.189Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T11:37:04.197282233Z 2020-06-02T11:37:04.197Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T11:37:04.419698687Z 2020-06-02T11:37:04.387Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T17:13:34.390024893Z 2020-06-02T17:13:34.389Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T20:57:42.896885748Z 2020-06-02T20:57:42.895Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T20:57:58.012645084Z 2020-06-02T20:57:57.993Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T20:58:03.903044028Z 2020/06/02 20:58:03 http: TLS handshake error from 10.154.0.89:44848: remote error: tls: bad certificate
2020-06-02T20:59:59.111683584Z 2020-06-02T20:59:59.111Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T20:59:59.451507454Z 2020-06-02T20:59:59.449Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:00:08.96931677Z 2020-06-02T21:00:08.969Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:00:08.972082809Z 2020-06-02T21:00:08.971Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:00:09.085292012Z 2020-06-02T21:00:09.083Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:00:09.088025244Z 2020-06-02T21:00:09.087Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:00:10.05514469Z 2020-06-02T21:00:10.052Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:00:26.386882258Z 2020/06/02 21:00:26 http: TLS handshake error from 10.154.0.3:42088: remote error: tls: bad certificate
2020-06-02T21:06:07.254377869Z 2020-06-02T21:06:07.252Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:16:04.834988005Z 2020-06-02T21:16:04.834Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:23:24.13032099Z 2020-06-02T21:23:24.129Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T21:23:31.127620578Z 2020-06-02T21:23:31.124Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T22:31:20.399379715Z 2020-06-02T22:31:20.397Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T22:34:33.014280792Z 2020-06-02T22:34:33.014Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T22:34:33.019973546Z 2020-06-02T22:34:33.019Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T22:34:33.071176991Z 2020-06-02T22:34:33.070Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T22:34:33.084315252Z 2020-06-02T22:34:33.084Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T22:34:35.328848206Z 2020-06-02T22:34:35.327Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T22:34:38.340745517Z 2020-06-02T22:34:38.340Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
2020-06-02T22:34:39.476182565Z 2020-06-02T22:34:39.467Z [INFO]  handler: Request received: Method=POST URL=/mutate?timeout=30s
```

**Comment by jasonodonnell** [comment]
Hi @therealsamlin, can you tell me more about your installation?

IE: did you mount certificates yourself or use the built in auto-tls?  Additionally did you scale the deployment of the injector beyond `replicas: 1`?

Thanks in advance.

**Comment by therealsamlin** [comment]
Hi @jasonodonnell, thanks for taking your time looking into this.

I've installed the injector manually, I did scale replicas to `replicas: 2` and add PDB to ensure that the vault-agent-injector is always running:

Here's the k8 config for my vault-agent-injector

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: vault-agent-injector
  namespace: default
  labels:
    app.kubernetes.io/name: vault-agent-injector
    app.kubernetes.io/instance: vault
    component: webhook
spec:
  replicas: 2
  selector:
    matchLabels:
      app.kubernetes.io/name: vault-agent-injector
      app.kubernetes.io/instance: vault
      component: webhook
  template:
    metadata:
      labels:
        app.kubernetes.io/name: vault-agent-injector
        app.kubernetes.io/instance: vault
        component: webhook
    spec:
      serviceAccountName: "vault-agent-injector"
      securityContext:
        runAsNonRoot: true
        runAsGroup: 1000
        runAsUser: 100
      containers:
        - name: sidecar-injector

          image: "hashicorp/vault-k8s:0.3.0"
          imagePullPolicy: "IfNotPresent"
          env:
            - name: AGENT_INJECT_LISTEN
              value: ":8080"
            - name: AGENT_INJECT_LOG_LEVEL
              value: info
            - name: AGENT_INJECT_VAULT_ADDR
              value: {vault_addr}
            - name: AGENT_INJECT_VAULT_IMAGE
              value: "vault:1.3.2"
            - name: AGENT_INJECT_TLS_AUTO
              value: vault-agent-injector-cfg
            - name: AGENT_INJECT_TLS_AUTO_HOSTS
              value: vault-agent-injector-svc,vault-agent-injector-svc.default,vault-agent-injector-svc.default.svc
              # replace auth path according to cluster name
            - name: AGENT_INJECT_VAULT_AUTH_PATH
              value: auth/au-rex-cluster  
          args:
            - agent-inject
            - 2>&1
          livenessProbe:
            httpGet:
              path: /health/ready
              port: 8080
              scheme: HTTPS
            failureThreshold: 2
            initialDelaySeconds: 1
            periodSeconds: 2
            successThreshold: 1
            timeoutSeconds: 5
          readinessProbe:
            httpGet:
              path: /health/ready
              port: 8080
              scheme: HTTPS
            failureThreshold: 2
            initialDelaySeconds: 2
            periodSeconds: 2
            successThreshold: 1
            timeoutSeconds: 5
```          

mutating webhook

```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: MutatingWebhookConfiguration
metadata:
  name: vault-agent-injector-cfg
  labels:
    app.kubernetes.io/name: vault-agent-injector
    app.kubernetes.io/instance: vault
webhooks:
  - name: vault.hashicorp.com
    clientConfig:
      service:
        name: vault-agent-injector-svc
        namespace: default
        path: "/mutate"
      caBundle:
    rules:
      - operations: ["CREATE", "UPDATE"]
        apiGroups: [""]
        apiVersions: ["v1"]
        resources: ["pods"]
```

SVC

```
apiVersion: v1
kind: Service
metadata:
  name: vault-agent-injector-svc
  namespace: default
  labels:
    app.kubernetes.io/name: vault-agent-injector
    app.kubernetes.io/instance: vault
spec:
  ports:
  - port: 443
    targetPort: 8080
  selector:
    app.kubernetes.io/name: vault-agent-injector
    app.kubernetes.io/instance: vault
    component: webhook
```

**Comment by jasonodonnell** [comment]
Hi @therealsamlin,

There's a bug with the auto-tls feature where deployments scaled beyond a `1` replica fight to update the `caBundle` for the mutating webhook.  They effectively go into an infinite loop trying to update the bundle.

If you scale down to 1 replica it will go away, or you could not use the auto-tls feature and manually mount TLS certificates for the webhook to use.  We're looking into a fix for this and it should be in the next release.

**Comment by therealsamlin** [comment]
Hi @jasonodonnell 

Do you have an example of how does one go about manually mounting TLS certificates? I'm using the `AGENT_INJECT_TLS_CERT_FILE` flag but am constantly getting this error:

```
2020-06-05T02:29:55.822939098Z Listening on ":8080"...
2020-06-05T02:29:55.822940072Z 2020-06-05T02:29:55.821Z [INFO]  handler: Starting handler..
2020-06-05T02:29:55.823582273Z Error listening: open : no such file or directory
```



my deployment config looks like this now:


```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/instance: vault
    app.kubernetes.io/name: vault-agent-injector
    component: webhook
  name: vault-agent-injector
  namespace: default
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/instance: vault
      app.kubernetes.io/name: vault-agent-injector
      component: webhook
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/instance: vault
        app.kubernetes.io/name: vault-agent-injector
        component: webhook
    spec:
      volumes:
      # vault cert (secret with key ca.crt)
      - name: vault-crt
        secret:
          secretName: vault-client-tls
      containers:
      - args:
        - agent-inject
        - 2>&1
        env:
        - name: AGENT_INJECT_LISTEN
          value: :8080
        - name: AGENT_INJECT_LOG_LEVEL
          value: info
        - name: AGENT_INJECT_VAULT_ADDR
          value: {vault_addr}
        - name: AGENT_INJECT_VAULT_IMAGE
          value: vault:1.3.2
        - name: AGENT_INJECT_TLS_CERT_FILE
          value: /etc/vault-crt/ca.crt  
        image: hashicorp/vault-k8s:0.2.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 2
          httpGet:
            path: /health/ready
            port: 8080
            scheme: HTTPS
          initialDelaySeconds: 1
          periodSeconds: 2
          successThreshold: 1
          timeoutSeconds: 5
        name: sidecar-injector
        volumeMounts:
          - name: vault-crt
            mountPath: /etc/vault-crt
        readinessProbe:
          failureThreshold: 2
          httpGet:
            path: /health/ready
            port: 8080
            scheme: HTTPS
          initialDelaySeconds: 2
          periodSeconds: 2
          successThreshold: 1
          timeoutSeconds: 5
```

Thank you!

**Comment by jasoares** [comment]
+1 for a fix for this

**Comment by mhonorio** [comment]
I have the same issue here.
Actually I think only 1 vault-agent-injector replica is not enough.

I'm having an issue during EKS node scaling down which lot of pods were stuck in CrashLoopBackOff because they don't find secrets. I'm suspecting that vault-agent-injector was also rescheduled during scaling and pods cannot grab secrets during rescheduling.

I'm using Helm to deploy Vault and we might rethink about replicas: 1. We should be able to increase this number or add PDB even transforming from Deployment to DaemonSet to guarantee we always have at least 1 agent-injector running.

**Comment by baldeynz** [comment]
+1 on a fix for this as well. We find that if we only use one pod then cluster upgrades or scale down events cause issues with other pods that depend on the injector, when the node that the injector is scheduled to disappears.

**Comment by ShabihZaidi** [comment]
@jasonodonnell anyone working on this? Can we help with this one if no one has picked so far

**Comment by fizzamahdi** [comment]
+1 for this fix

facing the same issue of wanting to have multiple replicas available + using a pod disruption budget so that pods using the vault sidecar do not go into a crashloopbackoff during node upgrades but getting a tls bad certificate error from vault agent injector when I do increase the replicas

**Comment by jasonodonnell** [comment]
Hi all, we're working on a fix for this, but wanted to give an update.  Likely the fix is going to land in Vault Helm to generate certificates for both the injector and Vault deployments out of the box.

While the auto-tls feature here is handy, it was lifted from sourced from another HashiCorp web hook (Consul Connect) and wasn't designed for multiple replicas.  This is evident in how the software creates a CA and signs certs.  We would rather not introduce state into the injector and there's no way to coordinate this without it (such as global locking).

This effort is likely going to launch with Vault 1.6. We're working on some features in the Vault project for bootstrapping TLS and we'll likely leverage that as our pre-start job.

## What Can I Do Today?

While auto-tls is the default feature, the injector does support user supplied certificates.  I've created a document with instructions on setting this up: https://www.vaultproject.io/docs/platform/k8s/helm/examples/injector-tls.  Using manual certs you can scale as fit!

If anyone is not using Helm to install the injector, please let me know and I can update the Vault K8s documentation for this use case.

**Comment by shuvankar0907** [comment]
> https://www.vaultproject.io/docs/platform/k8s/helm/examples/injector-tls

can you please provide a non helm version for injector installation.

**Comment by jbw815** [comment]
@shuvankar0907 you can use the "helm template" command with the --output-dir <path> to create a non-helm set of files.  You will want/need to create a values.yaml file with your specific values in it.  From that template you can then apply the K8s Yaml files with a kubectl apply to your environment.

**Comment by jasonodonnell** [comment]
This issue has been resolved in #198 and will be released in early January. Hope that helps!

**Comment by bholexy** [comment]
Awesome... Looking forward @jasonodonnell

**Comment by tvoran** [comment]
Hi folks, support for multiple injector replicas has been released with vault-k8s 0.7.0 and vault-helm 0.9.0, and we're adding documentation on the config options in this PR: https://github.com/hashicorp/vault/pull/10659, so I'll go ahead and close this issue for now.
